### ISSUE TITLE
threads_init not always present in gi bindings

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5790]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5790|
|Type|defect|
|Created|2012-07-17 18:14:57Z|

Jessica reported the following:

```
$ python -c "import twisted; print twisted.__version__; from twisted.internet import gtk3reactor"
 12.0.0+r34756
 Traceback (most recent call last):
   File "<string>", line 1, in <module>
   File "twisted/internet/gtk3reactor.py", line 22, in <module>
     from twisted.internet import gireactor
   File "twisted/internet/gireactor.py", line 53, in <module>
     _oldGiInit()
   File "twisted/internet/gireactor.py", line 42, in _oldGiInit
     GLib.threads_init()
   File "/usr/lib/python2.7/dist-packages/gi/module.py", line 268, in __getattr__
     return getattr(self._introspection_module, name)
   File "/usr/lib/python2.7/dist-packages/gi/module.py", line 101, in __getattr__
     self.__name__, name))
 AttributeError: 'gi.repository.GLib' object has no attribute 'threads_init'
```

<details><summary>Searchable metadata</summary>

```
trac-id__5790 5790
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__ 
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1342548897000000 1342548897000000
changetime__1345387705000000 1345387705000000
version__None None
owner__itamar itamar
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

What OS was this? What version of pygobject?

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess commented|
|-|-|

This is Ubuntu Natty:

```
$ apt-cache show python-gobject | grep Version
Version: 2.28.3-1ubuntu1
Version: 2.28.3-1ubuntu1.1
$ uname -a
Linux kid-charlemagne 2.6.38-13-generic #53-Ubuntu SMP Mon Nov 28 19:33:45 UTC 2011 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/issue
Ubuntu 11.04 \n \l
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

AFAICT it just doesn't exist in that version. `hasattr()` time!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [35156]) Branching to 'old-gi-5790'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

I will do this as part of [#5676](https://github.com/twisted/twisted/issues/5676), since otherwise the fixes will conflict.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35285]) Merge pygtkcompat-5676: Better support for newer and older gi versions.

Author: itamar
Review: jesstess
Fixes: [#5676](https://github.com/twisted/twisted/issues/5676), [#5790](https://github.com/twisted/twisted/issues/5790)

Work with old versions of gi that didn't wrap enable_threads(), and with newer versions that have the gi.pygtkcompat compatability layer.
